torch/lib/THD/base/data_channels/DataChannelNccl.hpp

```diff
@@ -0,0 +1,232 @@
+              THDGroup groupId = THDGroupWORLD) override;
+
+  void broadcast(std::vector<at::Tensor>& data,
+                 rank_type srcRank,
+                 THDGroup groupId = THDGroupWORLD) override;
+
+  void broadcast(at::Tensor& data,
+                 rank_type srcRank,
+                 THDGroup groupId = THDGroupWORLD) override;
+
+  void barrier(THDGroup groupId = THDGroupWORLD) override;
+
+  THDGroup newGroup(const std::vector<rank_type>& ranks) override;
+
+  void destroyGroup(THDGroup groupId = THDGroupWORLD) override;
+
+  // Not supported functions
+  void gather(std::vector<at::Tensor>& output,
+              at::Tensor& input,
+              rank_type dstRank,
+              THDGroup groupId = THDGroupWORLD) override;
+
+  void scatter(std::vector<at::Tensor>& input,
+               at::Tensor& output,
+               rank_type srcRank,
+               THDGroup groupId = THDGroupWORLD) override;
+
+  void send(Scalar& data, rank_type dstRank) override;
+
+  void send(at::Tensor& data, rank_type dstRank) override;
+
+  void receive(Scalar& data, rank_type srcRank) override;
+
+  rank_type receive(at::Tensor& data) override;
+
+  void receive(at::Tensor& data, rank_type srcRank) override;
+
+  RequestNccl* isend(at::Tensor& data, rank_type dstRank) override;
+
+  RequestNccl* ireceive(at::Tensor& data, rank_type srcRank) override;
+
+private:
+
+  // Current process' rank
+  rank_type _rank;
+  // Number of processes in network
+  rank_type _numProcesses;
+
+  // Accept waiting timeout in milliseconds, optional
+  int _timeout;
+  // Master's address
+  std::string _masterAddr;
+  // Master's port
+  port_type _masterPort;
+  // Socket on which the master is listening
+  int _masterListeningSocket;
+  // Sockets on which the master is sending to each slave
+  std::vector<int> _masterSendingSockets;
+  // Slave socket
+  int _slaveSocket;
+
+  // Number of GPUs on each node
+  int _numGPUs;
+  // Mutex for Nccl Data Channel
+  std::mutex _mutex;
+
+  /**
+   * The GPU devices each group have used. The GPU devices are stored in a
+   * device sequence.
+   *
+   * e.g. If the group only uses device 0, then the value of
+   *      the used device string stored (value of the hashmap) would be "0".
+   *
+   *      If the group uses device 0 - 7 and the each tensor of the
+   *      input tensor list is on device, 0, 1, 2, 3, 4, 5, 6, 7 separately,
+   *      then the value of the used device string stored would be
+   *      "0,1,2,3,4,5,6,7"
+   *
+   *      If the group uses device 0 - 7 and the each tensor of the
+   *      input tensor list is on device, 0, 4, 5, 6, 7, 1, 2, 3 separately,
+   *      then the value of the used device string  stored would be
+   *      "0,4,5,6,7,1,2,3"
+   *
+   *      Note that the order of the device for the tensor list matters and
+   *      each group can only be associated with one used device string
+   */
+  std::unordered_map<THDGroup, std::string> _groupDevices;
```

Comment:
The problem with this approach is that it doesn't really fit the interface we expose - a group is just a way to have multiple collectives running concurrently. I think a better solution, would be to store a communicator per group and per combination of devices used within this group. That would be equivalent and would satisfy the interface, without resorting to clearing caches, etc. It's not like this will create any more groups than if you didn't do this - it's just that certain communicators will belong to a single group instead of multiple.

I'm fine with merging this, but we definitely won't make this backend public until it's fixed.

Do we finally know what's causing the deadlocks and why do we even need to destroy communicators periodically?

Reply:
I don't know the exact reason why still but this is pretty much the fix for this symptom and that's something I would like to further investigate as well. Then I will probably open up a new PR to address this issue as well, by not associating a given process_group with just a set of GPU orders. I mean, I am not a big fan of caching all the communicators, I am thinking of caching probably the latest one, and also cache all the NCCL unique ID to avoid the broadcast. This is more like an optimization code and I will also do in a separate PR.

Reply:
Updates, now this is fixed and works. Instead of storing multiple communicators per group per device sequence. We only cache one communicator, if a new device sequence comes in, we destroy the previous communicator.  (The device sequence change shouldn't happen that often anyway in a training process)